### PR TITLE
make IRGen handle abstract inputs and attrs

### PR DESCRIPTION
**Pull request by marcrasi**
This makes IRGen able to handle abstract (i.e. non-deabstracted) inputs and attrs. I needed to do three things to achieve this:
1. Conform all our stdlib types to TensorGroup so that the TensorGroup case in `unpackAndAddInput` can handle them.
2. `unpackAndAddInput` had a bug where it didn't handle loadable types correctly in the TensorGroup case, so I fixed that and added a test that explicitly exercises loadable types in the TensorGroup case.
3. Make the attr cases use the underlying llvm type to determine the attr type, rather than the AST type, so that they "see through" all the Swift abstractions and just do the right thing based on the underlying value.

An alternative to (1) would have been to teach IRGen to extract the underlying TF values from all our stdlib types. I think we should do this eventually because it will generate faster code. But (1) is the easier thing to do right now, and the conformances are good to keep even after we implement the faster thing.

### CODE REVIEW

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/IRGen/IRGenSIL.cpp [hunk 1]
```diff
@@ -2079,22 +2082,37 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
       return llvm::ConstantInt::get(IGM.Int32Ty, 1);
     }
 
+    // Otherwise, this must conform to TensorGroup so we can add it using
+    // TFCOpAddInputFromTensorGroup.
```
Comment (mhong) [inline lib/IRGen/IRGenSIL.cpp]: typo: TFC_OpAddInputFromTensorGroup

File: stdlib/public/TensorFlow/TensorGroup.swift [hunk 2]
```diff
@@ -49,3 +51,82 @@ public extension TensorGroup {
...
+// Conform standard TensorFlow types to TensorGroup
+//===----------------------------------------------------------------------===//
+
+extension TensorHandle : TensorGroup where Scalar : AccelerableByTensorFlow {
+  public static var _typeList: [TensorDataType] {
+    return [Scalar.tensorFlowDataType]
+  }
+
+  public func _unpackTensorHandles(
+      into address: UnsafeMutablePointer<CTensorHandle>?) {
```
Comment (mhong) [inline stdlib/public/TensorFlow/TensorGroup.swift]: do we always force unwrap `address` via `!`? if so, should we pass in a non-optional `UnsafeMutablePointer<CTensorHandle>` instead?

**Review by mhong — APPROVED** [review]
nice!

File: stdlib/public/TensorFlow/TensorGroup.swift [hunk 3]
```diff
@@ -49,3 +51,82 @@ public extension TensorGroup {
     return Array(repeating: nil, count: Int(_tensorHandleCount))
   }
 }
+
+//===----------------------------------------------------------------------===//
+// Conform standard TensorFlow types to TensorGroup
+//===----------------------------------------------------------------------===//
+
+extension TensorHandle : TensorGroup where Scalar : AccelerableByTensorFlow {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/TensorGroup.swift]:
```suggestion
extension TensorHandle : TensorGroup {
```
I see no reason to add such a constraint. `Scalar` is already `AccelerableByTensorFlow` where it's bound.

Comment (marcrasi) [reply stdlib/public/TensorFlow/TensorGroup.swift]: done

File: lib/IRGen/IRGenSIL.cpp (hunk 1 above)
Comment (marcrasi) [reply lib/IRGen/IRGenSIL.cpp]: done

File: stdlib/public/TensorFlow/TensorGroup.swift (hunk 2 above)
Comment (marcrasi) [reply stdlib/public/TensorFlow/TensorGroup.swift]: `_TFCOpAddInputFromTensorGroup` passes it the `baseAddress` of an `UnsafeMutableBufferPointer`, which is an optional that is allowed to be nil when the buffer has 0 elements, which can happen for 0-element structs.

Comment (mhong) [reply stdlib/public/TensorFlow/TensorGroup.swift]: thanks. please document this case in https://github.com/apple/swift/blob/a4a9704b6019e44be7e0b4a0aa439a67b86f8561/stdlib/public/TensorFlow/TensorGroup.swift#L35. Same for `init()`?

Comment (mhong) [reply stdlib/public/TensorFlow/TensorGroup.swift]: Also, is there really such a thing as a 0-element struct?

Comment (rxwei) [reply stdlib/public/TensorFlow/TensorGroup.swift]: Yes, there is.

**Pull request merged**